feat(core): Custom session timeout and refresh configuration

despairblue opened the pull request:
## Summary
This allows configuring the session timeout and refresh via environment variables.

You can test it by starting n8n like this:

```sh
env N8N_USER_MANAGEMENT_JWT_DURATION_HOURS=0.005 N8N_USER_MANAGEMENT_JWT_REFRESH_TIMEOUT_HOURS=-1 pnpm dev
```

0.005 hours is 18 seconds and -1 means that the auto refresh of the cookie/JWT is disabled.
So you can log in, navigate around. Wait 18 seconds. Refresh and see that you're logged out.

If you want to test the auto refresh you can do this:
```sh
env N8N_USER_MANAGEMENT_JWT_DURATION_HOURS=0.005 N8N_USER_MANAGEMENT_JWT_REFRESH_TIMEOUT_HOURS=0.0025 pnpm dev
```

This will make the session last for 18 seconds and when less than 9 seconds are remaining it will issue a new cookie/JWT.

---

Reviewing this commit by commit can be easier. I mostly added tests before making changes, to make sure I don't break old behaviour.

## Related tickets and issues
https://linear.app/n8n/issue/PAY-1182/custom-session-timeout-configuration
https://linear.app/n8n/issue/DOC-734/add-custom-session-timeout-setting-to-environment-variables-page
https://github.com/n8n-io/n8n-docs/pull/1849


## Review / Merge checklist
- [x] PR title and summary are descriptive. **Remember, the title automatically goes into the changelog. Use `(no-changelog)` otherwise.** ([conventions](https://github.com/n8n-io/n8n/blob/master/.github/pull_request_title_conventions.md))
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included.
   > A bug is not considered fixed, unless a test is added to prevent it from happening again.
   > A feature is not complete without tests.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/n8n-io/n8n?pullRequest=8342) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/n8n-io/n8n?pullRequest=8342) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/n8n-io/n8n?pullRequest=8342) it.</sub>

despairblue committed:
add test for cli/auth/jwt

despairblue committed:
rename expiresIn and convert it to seconds

Rename `JwtToken.expiresIn` to `JwtToken.expiresInSeconds` to make it clear what unit it's supposed to be.
Also it's now seconds and not milliseconds. Reason for that is that seconds are a base unit of the SI, which makes it a good default. Also it's what the JWTs expect in `expireIn` and what the cookie Max-Age property expects.
Until now the function returned milliseconds and that was passed into the Max-Age property in `auth/jwt.issueCookie`. This caused no real issue, but it meant that the JWT would always expire before the cookie.

despairblue committed:
add units to the doc string

despairblue committed:
move mockInstance call to module root

no need to have it in the beforeEach

despairblue committed:
exporting `refreshExpiringCookie` to unit test it

Additionally I had to wrangle the types a bit. I want to call
`await refreshExpiringCookie(...)` in the unit tests, but to do this
without TS complaining the type needs account for the function being
async. Right now adding the type to the variable binding hides that
fact.
My workaround for that is to let TS infer the type of
`refreshExpiringCookie`. That lead to TS not being able to infer the
types of `res` and `next`. To fix that I used `satisfied`.

Happy to apply other solutions to this like extending the RequestHandler
type and creating an `AsyncRequestHandler` or adding an `Async` type
helper and using that.

```ts
type Async<T extends (...args: unknown[]) => unknown> = (
	...args: Parameters<T>
) => Promise<ReturnType<T>>;
```

despairblue committed:
adding unit test for middleware/auth.ts

despairblue committed:
remove TODO commment

despairblue committed:
use Time constant to convert between units of time

despairblue committed:
add jwtRefreshTimeoutHours and test that it resets to 0 if it's larger than jwtSessionDurationHours

despairblue committed:
clean up test

1. use Time constant for more readable time conversion
2. use jest-mock-extended to create users and requests
3. use type guards and jest.fail to remove unecessary assertions

despairblue committed:
make the refresh timeout configurable

despairblue committed:
remove comment

I'll add this as a comment in the review, where we can agree on where the warning should be emitted.

despairblue committed:
remove TODO comment

despairblue committed:
don't hard code default

despairblue committed:
don't use 2 different units in the test title and in the test's code

hours vs days

despairblue committed:
don't recalculate the same value multiple times

despairblue committed:
remove todo comment

netroy commented on packages/cli/src/Interfaces.ts:
```diff
@@ -667,7 +667,7 @@ export interface ILicensePostResponse extends ILicenseReadResponse {
 export interface JwtToken {
 	token: string;
-	expiresIn: number;
+	expiresInSeconds: number;
```
in the context of auth tokens, I've never seen an `expiresIn` that's not in seconds.
While I do see the additional clarity of mentioning seconds here, I don't think we need to rename the property itself.
I've seen an uptick in the number of long descriptive variable nameslately , and I fere like this makes code increasingly harder to read. 
IMO variable names should try to be as simple as possible, and all additional relevant metadata should move to JSDocs.

In this particular case, I'd prefer if renamed this back, and add a more descriptive JSDoc instead. That was anyone inspecting the code will still know that this is in seconds.

despairblue replied:
Happy to change it back, but some background before I do that.

> I've never seen an expiresIn that's not in seconds.

I actually thought exactly the same thing. So I assumed this is in seconds:
https://github.com/n8n-io/n8n/blob/b6d775768f927e69e7aa5f715c99f2fed4eaaa4c/packages/cli/src/auth/jwt.ts#L39-L41

And it is.

I also thought `maxAge` is in seconds:
https://github.com/n8n-io/n8n/blob/b6d775768f927e69e7aa5f715c99f2fed4eaaa4c/packages/cli/src/auth/jwt.ts#L88-L92

Because that's what it is in the browser: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#max-agenumber

And on master we currently pass milliseconds. So I thought I found a bug and fixed it in this PR. 

Now I tried it out locally and it turns out express expects milliseconds and converts it internally:
https://expressjs.com/en/api.html#res.cookie
![image](https://github.com/n8n-io/n8n/assets/927609/39e910b4-46f1-4505-8545-f864d3b9e0d0)

--- 

I'm happy to change it back, but I hope that shined some light on why I thought it may be helpful. If express would call their option `expiresInMilliseconds` I would have not introduced a bug in this PR 😄 

I'll fix the bug first that I have in the PR. Let me know what you think and if you still want me to revert the name change then I will.

krynble replied:
I was about to report that there was a bug in the duration, glad you've already figured it out.

netroy replied:
I get what you are saying and I agree that it's important to be clear about what unit a variable uses. My point was that we should use more JSDoc for that, and be leaner with our variable names. I've seen long and descriptive names get out of hand in old projects, and that makes code much harder to read because people put more details about what a variable does by putting all the details in the name.

We should instead [send PRs](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68255) (whenever possible) to add JSDoc on the packages we use to have the docs in our editors
![image](https://github.com/n8n-io/n8n/assets/196144/5cccb754-cf38-46f2-82f2-2c3334a8e7e9)

despairblue replied:
I agree. The JSDoc works well enough for me, so I changed the interface back and documented it via JSDoc.

Good thinking with amending the DT types for express. 
I hope your DT PR goes through 🤞🏾

despairblue committed:
fix: maxAge needs to be milliseconds

See: https://expressjs.com/en/api.html#res.cookie

krynble commented on packages/cli/src/middlewares/auth.ts:
```diff
@@ -41,17 +42,30 @@ const userManagementJwtAuth = (): RequestHandler => {
 		const cookieContents = jwt.decode(cookieAuth) as JwtPayload & { exp: number };
-		if (cookieContents.exp * 1000 - Date.now() < 259200000) {
+		if (cookieContents.exp * 1000 - Date.now() < jwtRefreshTimeoutMilliSeconds) {
 			// if cookie expires in < 3 days, renew it.
```
We can remove this line
```suggestion
```

despairblue replied:
good catch, I removed it

despairblue committed:
remove outdated comment

despairblue committed:
use JSDoc to document the unit instead of suffixing the variable name

despairblue committed:
add missing documentation

janober commented:
Got released with [`n8n@1.26.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.26.0)
